propolis/src/block/in_memory.rs

```diff
@@ -0,0 +1,279 @@
+                        // block if this thread goes into a quiesce state,
+                        // excluding all others from the queue lock.
+                        g.is_empty() && !sctx.pending_reqs()
+                    })
+                    .unwrap();
+            }
+        }
+    }
+
+    async fn do_scheduling(&self, actx: &AsyncCtx) {
+        loop {
+            let avail = self.idle_threads.acquire().await.unwrap();
+            avail.forget();
+
+            if let Some(req) = self.waiter.next(self.dev.as_ref(), actx).await {
+                let mut queue = self.queue.lock().unwrap();
+                queue.push_back(req);
+                drop(queue);
+                self.cv.notify_one();
+            }
+        }
+    }
+
+    fn spawn(self: &Arc<Self>, worker_count: NonZeroUsize, disp: &Dispatcher) {
+        for i in 0..worker_count.get() {
+            let tself = Arc::clone(self);
+
+            // Configure a waker to help threads to reach their yield points
+            // Doing this once (from thread 0) is adequate to wake them all.
+            let wake = if i == 0 {
+                let tnotify = Arc::downgrade(self);
+                Some(Box::new(move |_ctx: &DispCtx| {
+                    if let Some(this) = tnotify.upgrade() {
+                        let _guard = this.queue.lock().unwrap();
+                        this.cv.notify_all();
+                    }
+                }) as Box<WakeFn>)
+            } else {
+                None
+            };
+
+            let _ = disp
+                .spawn_sync(
+                    format!("file bdev {}", i),
+                    Box::new(move |mut sctx| {
+                        tself.blocking_loop(&mut sctx);
+                    }),
+                    wake,
+                )
+                .unwrap();
+        }
+
+        let sched_self = Arc::clone(self);
+        let actx = disp.async_ctx();
+        let sched_task = tokio::spawn(async move {
+            sched_self.do_scheduling(&actx).await;
+        });
+        // TODO: do we need to manipulate the task later?
+        disp.track(sched_task);
+    }
+}
+
+/// Read from bytes into guest memory
+fn process_read_request(
+    bytes: &Arc<Mutex<Vec<u8>>>,
+    offset: u64,
+    len: usize,
+    mappings: &Vec<SubMapping>,
+) -> Result<()> {
+    let bytes = bytes.lock().unwrap();
+    let data = &bytes[(offset as usize)..(offset as usize + len)];
+
+    let mut nwritten = 0;
+    for mapping in mappings {
+        let slice = &data[nwritten..(nwritten + mapping.len())];
+        let inner_nwritten = mapping.write_bytes(slice)?;
+
+        if inner_nwritten != mapping.len() {
+            return Err(std::io::Error::new(
+                ErrorKind::Other,
+                format!(
+                    "mapping.write_bytes failed! {} vs {}",
+                    inner_nwritten,
+                    mapping.len(),
+                ),
+            ));
+        }
```

Comment:
I assumed that `read_bytes` and `write_bytes` returned the number of bytes actually performed (similar to  std::io::Read::read) because of the return type of `Result<usize>`, but looking at the code it always returns the number of bytes of the buffer. In memory and crucible changed in 9a0ca32